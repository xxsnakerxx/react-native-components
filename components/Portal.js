/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */

import React from 'react';

import {
  StyleSheet,
  View,
} from 'react-native';

let _portalRef;
let lastUsedTag = 0;

export default class Portal extends React.Component {
  // eslint-disable-next-line no-plusplus
  static allocateTag = () => `__modal_${++lastUsedTag}`;

  static showModal = (tag, component) => {
    if (!_portalRef) {
      throw new Error('Calling showModal but no Portal has been rendered.');
    }

    _portalRef._showModal(tag, component);
  }

  static closeModal = (tag) => {
    if (!_portalRef) {
      throw new Error('Calling closeModal but no Portal has been rendered.');
    }

    _portalRef._closeModal(tag);
  }

  static getOpenModals = () => {
    if (!_portalRef) {
      throw new Error('Calling getOpenModals but no Portal has been rendered.');
    }

    return _portalRef._getOpenModals();
  }

  constructor(props) {
    super(props);

    this.state = {
      modals: {},
    };
  }

  _showModal(tag, component) {
    const {
      modals,
    } = this.state;

    this.setState({
      modals: {
        ...modals,
        [tag]: component,
      },
    });
  }

  _closeModal(tag) {
    const {
      modals,
    } = this.state;

    const modalsCopy = { ...modals };

    if (!modalsCopy[tag]) {
      return;
    }

    delete modalsCopy[tag];

    this.setState({
      modals: modalsCopy,
    });
  }

  _getOpenModals() {
    const {
      modals,
    } = this.state;

    return Object.keys(modals);
  }

  render() {
    const {
      modals,
    } = this.state;

    _portalRef = this;

    if (!Object.keys(modals).length) {
      return null;
    }

    return (
      <View
        style={styles.modalsContainer}
        pointerEvents="box-none"
      >
        {Object.keys(modals).map((tag) => ({ ...modals[tag], key: tag }))}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  modalsContainer: { ...StyleSheet.absoluteFillObject },
});
