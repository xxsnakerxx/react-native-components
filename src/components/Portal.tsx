/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */

import React, {ReactElement} from 'react';

import {StyleSheet, View} from 'react-native';

let portalRef: Portal | null;

let counter = 0;

interface State {
  modals: {[key: string]: ReactElement};
}

export default class Portal extends React.PureComponent<any, State> {
  static allocateTag = () => `__modal_${++counter}`;

  static showModal = (tag: string, component: ReactElement<any>) => {
    if (!portalRef) {
      if (process.env.NODE_ENV === 'development') {
        console.warn('Calling showModal but no Portal has been rendered.');
      }

      return;
    }

    portalRef._showModal(tag, component);
  };

  static closeModal = (tag: string) => {
    if (!portalRef) {
      if (process.env.NODE_ENV === 'development') {
        console.warn('Calling closeModal but no Portal has been rendered.');
      }

      return;
    }

    portalRef._closeModal(tag);
  };

  static getOpenModals = () => {
    if (!portalRef) {
      if (process.env.NODE_ENV === 'development') {
        console.warn('Calling getOpenModals but no Portal has been rendered.');
      }

      return [];
    }

    return portalRef._getOpenModals();
  };

  constructor(props: any) {
    super(props);

    this.state = {
      modals: {},
    };
  }

  componentDidMount() {
    // eslint-disable-next-line consistent-this
    portalRef = this;
  }

  componentWillUnmount() {
    portalRef = null;
    counter = 0;
  }

  _showModal(tag: string, component: ReactElement) {
    const {modals} = this.state;

    this.setState({
      modals: {
        ...modals,
        [tag]: component,
      },
    });
  }

  _closeModal(tag: string) {
    const {modals} = this.state;

    const modalsCopy = {...modals};

    if (!modalsCopy[tag]) {
      return;
    }

    delete modalsCopy[tag];

    this.setState({
      modals: modalsCopy,
    });
  }

  _getOpenModals() {
    const {modals} = this.state;

    return Object.keys(modals);
  }

  render() {
    const {modals} = this.state;

    if (!Object.keys(modals).length) {
      return null;
    }

    return (
      <View style={styles.modalsContainer} pointerEvents="box-none">
        {Object.keys(modals).map((tag) => ({...modals[tag], key: tag}))}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  modalsContainer: {...StyleSheet.absoluteFillObject},
});
