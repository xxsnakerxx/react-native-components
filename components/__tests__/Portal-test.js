import {
  Text,
} from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Portal from '../Portal';

describe('Portal', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <Portal />,
    );

    Portal.showModal(
      Portal.allocateTag(),
      (
        <Text>
          Modal 1
        </Text>
      ),
    );
    Portal.showModal(
      Portal.allocateTag(),
      (
        <Text>
          Modal 2
        </Text>
      ),
    );

    expect(component.toJSON()).toMatchSnapshot();
    expect(Portal.getOpenModals()).toMatchSnapshot();
  });
});
