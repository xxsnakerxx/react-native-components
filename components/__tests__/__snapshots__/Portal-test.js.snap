// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <Text>
    Modal 1
  </Text>
  <Text>
    Modal 2
  </Text>
</View>
`;

exports[`Portal renders correctly 2`] = `
Array [
  "__modal_1",
  "__modal_2",
]
`;
